javalib/src/main/scala/java/time/Duration.scala

```diff
@@ -0,0 +1,271 @@
+    new Duration(newSeconds, newNanos)
+  }
+
+  def dividedBy(divisor: Long): Duration = {
+    val secondsQuot = normalizedSeconds / divisor
+    val secondsRem = normalizedSeconds % divisor
+    val nanos = {
+      try {
+        val total = Math.addExact(Math.multiplyExact(secondsRem, NANOS_IN_SECOND),
+            normalizedNanos)
+        total / divisor
+      } catch {
+        case _: ArithmeticException =>
+          val total = BigInt(secondsRem) * NANOS_IN_SECOND + normalizedNanos
+          (total / divisor).toLong
+      }
+    }
+    Duration.ofSeconds(secondsQuot).plusNanos(nanos)
+  }
+
+  def negated(): Duration = multipliedBy(-1)
+
+  def abs(): Duration = if (isNegative()) negated() else this
+
+  def addTo(temporal: Temporal): Temporal =
+    temporal.plus(normalizedSeconds, SECONDS).plus(normalizedNanos, NANOS)
+
+  def subtractFrom(temporal: Temporal): Temporal =
+    temporal.minus(normalizedSeconds, SECONDS).minus(normalizedNanos, NANOS)
+
+  def toDays(): Long = seconds / SECONDS_IN_DAY
+
+  def toHours(): Long = seconds / SECONDS_IN_HOUR
+
+  def toMinutes(): Long = seconds / SECONDS_IN_MINUTE
+
+  def toMillis(): Long = {
+    val millis1 = Math.multiplyExact(seconds, MILLIS_IN_SECOND)
+    val millis2 = nanos / NANOS_IN_MILLI
+    Math.addExact(millis1, millis2)
+  }
+
+  def toNanos(): Long =
+    Math.addExact(Math.multiplyExact(seconds, NANOS_IN_SECOND), nanos)
+
+  def compareTo(that: Duration): Int = {
+    val secCmp = seconds.compareTo(that.getSeconds)
+    if (secCmp == 0) nanos.compareTo(that.getNano)
+    else secCmp
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Duration =>
+      seconds == that.getSeconds && nanos == that.getNano
+
+    case _ => false
+  }
+
+  override def hashCode(): Int = 31 * seconds.hashCode + nanos
+
+  override def toString(): String = {
+    val mins = normalizedSeconds / 60
+    val secsOfMin = normalizedSeconds % 60
+    val hours = mins / 60
+    val minsOfHour = mins % 60
+    val hourPart = if (hours == 0) "" else hours.toString + "H"
+    val minPart = if (minsOfHour == 0) "" else minsOfHour.toString + "M"
+    val nanos1 = math.abs(normalizedNanos)
+    val decimals = f"$nanos1%09d".reverse.dropWhile(_ == '0').reverse
+    val decimalPart = if (decimals.isEmpty) "" else "." + decimals
+    val secsPart = secsOfMin match {
+      case 0 if seconds != 0 && nanos == 0 => ""
+      case 0 if seconds < 0                => "-0" + decimalPart + "S"
+      case n                               => n.toString + decimalPart + "S"
+    }
+    "PT" + hourPart + minPart + secsPart
+  }
+}
+
+object Duration {
+  import Constants._
+
+  final val ZERO = new Duration(0, 0)
+
+  private final val MIN = new Duration(Long.MinValue, 0)
+
+  private final val MAX = new Duration(Long.MaxValue, 999999999)
+
+  def ofDays(days: Long): Duration = // of(days, ChronoUnit.DAYS)
+    ofSeconds(SECONDS_IN_DAY).multipliedBy(days)
```

Comment:
If I use `of(days, ChronoUnit.DAYS)` many tests are failing, because `ChronoUnit.DAYS` sometimes evaluates to `null`. This looks like a bug to me. Could one of you investigate?